src/server/api/lib/twilio.js

```diff
@@ -194,72 +201,90 @@ async function sendMessage(message, contact, trx) {
     }
 
     twilio.messages.create(messageParams, (err, response) => {
-      const messageToSave = {
-        ...message
-      };
-      log.info("messageToSave", messageToSave);
-      let hasError = false;
-      if (err) {
-        hasError = true;
-        log.error("Error sending message", err);
-        console.log("Error sending message", err);
-        messageToSave.service_response += JSON.stringify(err);
-      }
-      if (response) {
-        messageToSave.service_id = response.sid;
-        hasError = !!response.error_code;
-        messageToSave.service_response += JSON.stringify(response);
-      }
+      postMessageSend(message, contact, resolve, reject, err, response);
+    });
+  });
+}
 
-      if (hasError) {
-        const SENT_STRING = '"status"'; // will appear in responses
-        if (
-          messageToSave.service_response.split(SENT_STRING).length >=
-          MAX_SEND_ATTEMPTS + 1
-        ) {
-          messageToSave.send_status = "ERROR";
-        }
-        let options = { conflict: "update" };
-        if (trx) {
-          options.transaction = trx;
-        }
-        Message.save(messageToSave, options)
-          // eslint-disable-next-line no-unused-vars
-          .then((_, newMessage) => {
-            reject(
-              err ||
-                (response
-                  ? new Error(JSON.stringify(response))
-                  : new Error("Encountered unknown error"))
-            );
-          });
-      } else {
-        let options = { conflict: "update" };
-        if (trx) {
-          options.transaction = trx;
-        }
-        Message.save(
-          {
-            ...messageToSave,
-            send_status: "SENT",
-            service: "twilio",
-            sent_at: new Date()
-          },
-          options
-        ).then((saveError, newMessage) => {
-          resolve(newMessage);
-        });
+function postMessageSend(message, contact, resolve, reject, err, response) {
+  const messageToSave = {
+    ...message
+  };
+  log.info("messageToSave", messageToSave);
+  let hasError = false;
+  if (err) {
+    hasError = true;
+    log.error("Error sending message", err);
+    console.log("Error sending message", err);
+  }
+  if (response) {
+    messageToSave.service_id = response.sid;
+    hasError = !!response.error_code;
+  }
+
+  if (hasError) {
+    if (
+      messageToSave.error_code < -MAX_SEND_ATTEMPTS &&
+      messageToSave.error_code > -100
+    ) {
+      messageToSave.send_status = "ERROR";
+    }
+    // decrement error code starting from zero
+    messageToSave.error_code = Number(response.error_code || 0) - 1;
+    let contactUpdateQuery = Promise.resolve(1);
```

Comment:
`postMessageSend` is mostly just pulled out of `sendMessage` for testing purposes.  There are a couple changes:
1. message.service_response refs are removed
2. message.error_code references are added along with logic that accomodates a similar use of service_response which sees how many times we try to send before giving up.
3. `contactUpdateQuery` updates campaign_contact.error_code with the error state.